blink/legacy.c

```diff
@@ -68,53 +82,64 @@ static relegated void Pushad(P) {
   memcpy(b[5], m->dx, 4);
   memcpy(b[6], m->cx, 4);
   memcpy(b[7], m->ax, 4);
-  Put64(m->sp, (v = (Get32(m->sp) - sizeof(b)) & 0xffffffff));
-  CopyToUser(m, m->ss.base + v, b, sizeof(b));
+  PushaCommon(m, b, sizeof(b));
 }
 
-static relegated void Popaw(P) {
+static relegated void PopaCommon(struct Machine *m, void *b, size_t n) {
   u64 addr;
-  u8 b[8][2];
-  addr = m->ss.base + Read16(m->sp);
-  if (CopyFromUser(m, b, addr, sizeof(b)) == -1) {
-    ThrowSegmentationFault(m, addr);
+  switch (m->mode.omode) {
+    case XED_MODE_REAL:
+      addr = m->ss.base + Read16(m->sp);
+      if (CopyFromUser(m, b, addr, n) == -1) {
+        ThrowSegmentationFault(m, addr);
+      }
+      Put16(m->sp, (Get16(m->sp) + n) & 0xffff);
+      break;
+    case XED_MODE_LEGACY:
+      addr = m->ss.base + Read32(m->sp);
+      if (CopyFromUser(m, b, addr, n) == -1) {
+        ThrowSegmentationFault(m, addr);
+      }
+      Put64(m->sp, (Get32(m->sp) + n) & 0xffffffff);
+      break;
+    default:
+      __builtin_unreachable();
   }
-  Put16(m->sp, (Get32(m->sp) + sizeof(b)) & 0xffff);
+}
+
+static relegated void Popaw(P) {
+  u8 b[8][2];
+  PopaCommon(m, b, sizeof(b));
   memcpy(m->di, b[0], 2);
   memcpy(m->si, b[1], 2);
   memcpy(m->bp, b[2], 2);
-  memcpy(m->sp, b[3], 2);
   memcpy(m->bx, b[4], 2);
   memcpy(m->dx, b[5], 2);
   memcpy(m->cx, b[6], 2);
   memcpy(m->ax, b[7], 2);
 }
 
 static relegated void Popad(P) {
-  u64 addr;
   u8 b[8][4];
-  addr = m->ss.base + Get32(m->sp);
-  if (CopyFromUser(m, b, addr, sizeof(b)) == -1) {
-    ThrowSegmentationFault(m, addr);
-  }
-  Put64(m->sp, (Get32(m->sp) + sizeof(b)) & 0xffffffff);
+  PopaCommon(m, b, sizeof(b));
   memcpy(m->di, b[0], 4);
   memcpy(m->si, b[1], 4);
   memcpy(m->bp, b[2], 4);
-  memcpy(m->sp, b[3], 4);
   memcpy(m->bx, b[4], 4);
   memcpy(m->dx, b[5], 4);
   memcpy(m->cx, b[6], 4);
   memcpy(m->ax, b[7], 4);
 }
 
 relegated void OpPusha(P) {
-  switch (Eamode(rde)) {
+  switch (m->mode.omode) {
```

Comment:
Why aren't you using `Eamode()`? Don't you want to support the operand-size override prefix?

Reply:
Hello @jart,

I found that `Eamode()` is not realy the correct property to use here &mdash; specifically, it is controlled by the "address size" prefix (`0x67`) rather than the "operand size" prefix (`0x66`).

Thank you!

Reply:
I believe using `kStackOsz[]` to determine the operand size will (also) do the correct thing.  But `kStackOsz[]` happens to be in `stack.c` and not `legacy.c`.

Thank you!